switch/traceext.ml

```diff
@@ -16,23 +16,139 @@
 
 open Lwt
 
-let size = 128
+type message_id = string * int64
 
-let buffer : (int64 * Message_switch_core.Protocol.Event.t) option array =
-  Array.make size None
+module CompactMessage : sig
+  type t
+
+  val of_message : Message_switch_core.Protocol.Message.t -> t
+
+  val to_message : t -> Message_switch_core.Protocol.Message.t
+
+  val truncate_at : int ref
+end = struct
+  open Message_switch_core.Protocol
+
+  let truncate_at = ref 0
+
+  type kind = Message.kind
+
+  type payload = Raw of string | Rpc of Rpc.t
+
+  type t = {payload: payload; kind: Message.kind}
+
+  let to_message t =
+    {
+      Message.payload=
+        (match t.payload with Rpc rpc -> Jsonrpc.to_string rpc | Raw s -> s)
+    ; kind= t.kind
+    }
+
+  let truncated = "…"
```

Comment:
This is a UTF-8 encoded glyph. We could stick with ASCII and use "..." but I understand it's the 21st century now.